models/siegert_neuron.cpp

```diff
@@ -0,0 +1,479 @@
+  const size_t quantity = kernel().connection_manager.get_min_delay();
+  B_.drift_input_.resize( quantity, 0.0 );
+  B_.diffusion_input_.resize( quantity, 0.0 );
+  B_.last_y_values.resize( quantity, 0.0 );
+
+  B_.logger_.reset(); // includes resize
+  Archiving_Node::clear_history();
+}
+
+void
+nest::siegert_neuron::calibrate()
+{
+  B_.logger_
+    .init(); // ensures initialization in case mm connected after Simulate
+
+  const double h = Time::get_resolution().get_ms();
+
+  // propagators
+  V_.P1_ = std::exp( -h / P_.tau_ );
+  V_.P2_ = -numerics::expm1( -h / P_.tau_ );
+}
+
+/* ----------------------------------------------------------------
+ * Update and event handling functions
+ */
+
+bool
+nest::siegert_neuron::update_( Time const& origin,
+  const long from,
+  const long to,
+  const bool wfr_update )
+{
+  assert(
+    to >= 0 && ( delay ) from < kernel().connection_manager.get_min_delay() );
+  assert( from < to );
+
+  bool done = true;
+  const size_t quantity = kernel().connection_manager.get_min_delay();
+  const double wfr_tol = kernel().simulation_manager.get_wfr_tol();
+
+  // allocate memory to store rates to be sent by rate events
+  std::vector< double > new_rates( quantity, 0.0 );
+
+  for ( long lag = from; lag < to; ++lag )
+  {
+    // register rate in buffer
+    new_rates[ lag ] = S_.r_;
+
+    // propagate rate to new time step (exponential integration)
+    double drive = siegert( P_.tau_m_ * 1e-3,
+      P_.tau_syn_ * 1e-3,
+      P_.t_ref_ * 1e-3,
+      P_.theta_,
+      P_.V_reset_,
+      B_.drift_input_[ lag ],
+      std::sqrt( B_.diffusion_input_[ lag ] ) );
+    S_.r_ = V_.P1_ * ( S_.r_ ) + ( 1 - V_.P1_ ) * P_.mean_ + V_.P2_ * drive;
+
+    if ( not wfr_update )
+    {
+      // rate logging
+      B_.logger_.record_data( origin.get_steps() + lag );
+    }
+
+    if ( wfr_update ) // check convergence of waveform relaxation
+    {
+      done = ( fabs( S_.r_ - B_.last_y_values[ lag ] ) <= wfr_tol ) && done;
+      // update last_y_values for next wfr_update iteration
+      B_.last_y_values[ lag ] = S_.r_;
+    }
+  }
+
+  if ( not wfr_update )
+  {
+    // clear last_y_values
+    B_.last_y_values.clear();
+    B_.last_y_values.resize( quantity, 0.0 );
+
+    // modifiy new_rates for diffusion-event as proxy for next min_delay
+    for ( long temp = from; temp < to; ++temp )
+      new_rates[ temp ] = S_.r_;
+  }
+
+  // Send diffusion-event
+  DiffusionEvent rve;
+  rve.set_coeffarray( new_rates );
+  kernel().event_delivery_manager.send_secondary( *this, rve );
+
+  // Reset variables
+  B_.drift_input_.clear();
+  B_.drift_input_.resize( quantity, 0.0 );
+  B_.diffusion_input_.clear();
+  B_.diffusion_input_.resize( quantity, 0.0 );
+
+
+  return done;
+}
+
+void
+nest::siegert_neuron::handle( DiffusionEvent& e )
+{
+  size_t i = 0;
+  std::vector< unsigned int >::iterator it = e.begin();
+  // The call to get_coeffvalue( it ) in this loop also advances the iterator it
+  while ( it != e.end() )
```

Comment:
@heplesser I agree with you, that the writing into the communication buffer and the reading from the communication buffer of secondary events is only safe if the functions `read_from_communication_buffer` and `write_to_communication_buffer` respectively the operators `operator<<` and `operator>>` match and do not contain any bugs. This is crucial here, otherwise we almost certainly add up with a segmentation fault. But if this is done correctly, then this code here is perfectly safe (although I see that this is not obvious):

All we do here is reading from the double vector that is included in the corresponding event. Since we read it directly from the receive buffer to avoid a copy of the data it is still saved as a unsigned int vector. However this double vector has a predefined length `coeff_length_`. `coeff_length_` is a static member variable to each event class and is set only once in `NodeManager::check_wfr_use()`. An `assert` in `void set_coeffarray` ensures that every event that is send has a double vector of the right length. As `e.begin()` and `e.end()` are iterators pointing to the beginning and end of the double vector - we do exactly know that we cannot exceed the size of `B_.drift_input_` (because it is the same as `coeff_length_` of the event).

We could add a `assert( e.coeff_length_ == B_.drift_input_.size() )` before the while-loop here if you think this would be helpful?

The same arguments hold for the different event classes and handle functions.

Regarding your last bullet point: Is there any benefit in moving the `while` loop to the event class? If there is I do not see it at the moment...